skbio/io/registry.py

```diff
@@ -514,24 +519,37 @@ def _read_gen(self, file, fmt, into, verify, kwargs):
         # kwargs should still retain the contents of io_kwargs because the
         # actual reader will also need them.
         with _resolve_file(file, **io_kwargs) as (file, _, _):
-            reader, kwargs = self._init_reader(
+            reader, kwargs, consumed = self._init_reader(
                 file, fmt, into, verify, kwargs, io_kwargs
             )
+            # Add consumed lines to the start of the file to "reset" it
+            if consumed is not None and not file.seekable():
+                file = itertools.chain(consumed, file)
             yield from reader(file, **kwargs)
 
     def _find_io_kwargs(self, kwargs):
         return {k: kwargs[k] for k in _open_kwargs if k in kwargs}
 
     def _init_reader(self, file, fmt, into, verify, kwargs, io_kwargs):
         skwargs = {}
+        consumed = []
+        # Get sniffer from format (defined in iosources)
         if fmt is None:
             fmt, skwargs = self.sniff(file, **io_kwargs)
         elif verify:
             sniffer = self.get_sniffer(fmt)
             if sniffer is not None:
-                backup = file.tell()
-                is_format, skwargs = sniffer(file, **io_kwargs)
-                file.seek(backup)
+                try:
+                    # Attempt to reset and sniff file
+                    backup = file.tell()
+                    res = sniffer(file, **io_kwargs)
+                    is_format = res[0]
+                    skwargs = res[1]
+                    file.seek(backup)
+                except io.UnsupportedOperation:
+                    # File is not seekable, so get consumed lines to reset it
+                    is_format, skwargs, consumed = sniffer(file, **io_kwargs)
```

Comment:
`peek` for `stdin` (see `sys.stdin.buffer.peek`) and many other IO types (not `io.StringIO` AFAIK) would work without advancing the pointer too. But, it's read request is not assured to be what you want in terms of bytes

Reply:
Yeah, I've tested a couple other methods of reading nonseekables including this but it just seems that reconstructing it as a chain is the best.